src/main/java/org/cloudfoundry/promregator/lite/cfaccessor/SimpleCFAccessorImpl.kt

```diff
@@ -0,0 +1,511 @@
+            }
+            query { applicationsV3().list(it) }
+        }
+    }
+
+    private suspend fun <R: ResourceV2<*>, T: org.cloudfoundry.client.v2.PaginatedResponse<R>,S : PaginatedRequestV2> cfQueryV2(api: String, type: RequestType, requestBlock: CfClientDslV2<R, T, S>.() -> Unit): List<R> {
+        val cfClient = cloudFoundryClients[api]
+        val dsl = CfClientDslV2<R, T, S>(cfClient, type)
+        dsl.requestBlock()
+        return dsl.runQueries()
+    }
+
+    private suspend fun <R: Resource, T: org.cloudfoundry.client.v3.PaginatedResponse<R>,S : V3PaginatedRequest> cfQueryV3(api: String, type: RequestType, requestBlock: CfClientDslV3<R, T, S>.() -> Unit): List<R> {
+        val cfClient = cloudFoundryClients[api]
+        val dsl = CfClientDslV3<R, T, S>(cfClient, type)
+        dsl.requestBlock()
+        return dsl.runQueries()
+    }
+
+    inner class CfClientDslV2<R: ResourceV2<*>, T: org.cloudfoundry.client.v2.PaginatedResponse<R>, S: PaginatedRequestV2>(
+            private val cfClient: ReactorCloudFoundryClient?,
+            private val type: RequestType,
+            ) {
+        lateinit var requestTemplate : (Int) -> S
+        lateinit var queryTemplate: ReactorCloudFoundryClient.(S) -> Mono<T>
+
+        fun request(block: (Int) -> S) {
+            requestTemplate = block
+        }
+
+        fun query(block: ReactorCloudFoundryClient.(S) -> Mono<T>) {
+            queryTemplate = block
+        }
+
+        suspend fun runQueries() : List<R> {
+            val req = requestTemplate(1)
+
+            val clientSafe = cfClient ?: return listOf()
+            val first = rateLimitAndRecord(type) {clientSafe.queryTemplate(req).awaitSingle() }
+            val totalPages = first?.totalPages ?: 0
+            val firstResources = first?.resources ?: listOf()
+
+            val otherPages = if(totalPages > 1) {
+                coroutineScope {
+                    (1..totalPages).map {
+                        val otherReq = requestTemplate(it)
+                        async { rateLimitAndRecord(type) { clientSafe.queryTemplate(otherReq).awaitSingle() } }
+                    }.awaitAll().flatMap { it?.resources ?: listOf() }
+                }
+
+            } else {
+                listOf()
+            }
+
+            return (firstResources+otherPages)
+        }
+    }
+
+
+    inner class CfClientDslV3<R: Resource, T: org.cloudfoundry.client.v3.PaginatedResponse<R>, S: V3PaginatedRequest>(
+            private val cfClient: ReactorCloudFoundryClient?,
+            private val type: RequestType,
+            ) {
+        lateinit var requestTemplate : (Int) -> S
+        lateinit var queryTemplate: ReactorCloudFoundryClient.(S) -> Mono<T>
+
+        fun request(block: (Int) -> S) {
+            requestTemplate = block
+        }
+
+        fun query(block: ReactorCloudFoundryClient.(S) -> Mono<T>) {
+            queryTemplate = block
+        }
+
+        suspend fun runQueries() : List<R> {
```

Comment:
I had to create a v3 version of the same logic since it uses a different pagination class